src/basic/smack-util.c

```diff
@@ -45,84 +45,162 @@ bool mac_smack_use(void) {
 #endif
 }
 
-int mac_smack_apply(const char *path, const char *label) {
-        int r = 0;
+static int mac_get_xattr(const char *path, const char *attr, const char **label) {
+        char l[NAME_MAX] = "", *ll = NULL;
+        ssize_t s = 0;
 
         assert(path);
+        assert(attr);
+        assert(label);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = lsetxattr(path, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = lremovexattr(path, "security.SMACK64");
-        if (r < 0)
+        s = getxattr(path, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
+
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
+
+        *label = ll;
 #endif
 
-        return r;
+        return s;
 }
 
-int mac_smack_apply_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_get_xattr_fd(int fd, const char *attr, const char **label) {
+        char l[NAME_MAX] = "", *ll = NULL;
+        ssize_t s = 0;
 
         assert(fd >= 0);
+        assert(attr);
+        assert(label);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = fsetxattr(fd, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = fremovexattr(fd, "security.SMACK64");
-        if (r < 0)
+        s = fgetxattr(fd, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
+
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
+
+        *label = ll;
 #endif
 
-        return r;
+        return s;
 }
 
-int mac_smack_apply_ip_out_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_set_xattr(const char *path, const char *attr, const char *label) {
+        int r;
 
-        assert(fd >= 0);
+        assert(path);
+        assert(attr);
 
 #ifdef HAVE_SMACK
```

Comment:
Better, but please move the `#ifdef` up so it stubs out the *entire* function, and then make an `#else` branch with

    static int mac_set_xattr(const char *path, const char *attr, const char *label) {
        return 0;
    }

Both functions can then live under the same `#ifdef` guard.